ingester/src/server/grpc/query/instrumentation.rs

```diff
@@ -49,17 +52,25 @@ pub(crate) struct FlightFrameEncodeRecorder {
     inner: FlightDataEncoder,
     parent_span: Option<Span>,
     state: FrameEncodeRecorderState,
+    encoding_duration_metric: Arc<DurationHistogram>,
+    current_duration_subtotal: DurationCounter,
 }
 
 impl FlightFrameEncodeRecorder {
     /// Creates a [`FlightFrameEncodeRecorder`].
     ///
     /// When a `parent_span` is provided, it adds a trace span as a child of the parent.
-    pub fn new(inner: FlightDataEncoder, parent_span: Option<Span>) -> Self {
+    pub fn new(
+        inner: FlightDataEncoder,
+        parent_span: Option<Span>,
+        encoding_duration_metric: Arc<DurationHistogram>,
+    ) -> Self {
         Self {
             inner,
             parent_span,
             state: FrameEncodeRecorderState::Unpolled,
+            encoding_duration_metric,
+            current_duration_subtotal: DurationCounter::default(),
```

Comment:
This should be just a [`Duration`](https://doc.rust-lang.org/std/time/struct.Duration.html):

```suggestion
            current_duration_subtotal: Duration::ZERO,
```

You then later can do `current_duration_subtotal += measured_difference` below.

We shouldn't use the metric types as general purpose primitives! I am pretty sure we discussed this before, but the metrics are specialised types that have no cross-thread synchronisation. There's no guarantee that the `fetch()` you do later is going to have the right value as there's no happens-before semantics being upheld (any async code runs across multiple threads). Beyond that, it's just repurposing something that's not intended for this purpose, which comes with overhead too :+1:

Reply:
done